### PR TITLE
`z_parameter`: Magic

### PR DESCRIPTION
**Pull request by engineer124**
The magic system is slightly changed compared to OoT, specifically they decided to separate adding magic and consuming magic into their own independent system/functions. 

They also added some functionality for MM specific cases (deku bubble, goron spikes, zora shocks, fierce deity beams, giants mask usage)

### CODE REVIEW

File: src/code/z_parameter.c [hunk 1]
```diff
@@ -1145,26 +1149,474 @@ void Inventory_ChangeAmmo(s16 item, s16 ammoChange) {
...
+            break;
+
+        case MAGIC_STATE_CONSUME_LENS:
+            // Slowly consume magic while lens is on
+            if ((play->pauseCtx.state == 0) && (play->pauseCtx.debugState == 0) && (msgCtx->msgMode == 0) &&
+                (play->gameOverCtx.state == 0) && (play->sceneLoadFlag == 0) && (play->unk_18B4A == 0) &&
+                !Play_InCsMode(play)) {
+
+                if ((gSaveContext.save.playerData.magic == 0) ||
+                    ((func_801242DC(play) >= 2) && (func_801242DC(play) < 5)) ||
```
Comment (EllipticEllipsis) [inline src/code/z_parameter.c:1464]: <= might make more sense here, but can't really tell for now unfortunately.

File: src/code/z_parameter.c [hunk 2]
```diff
@@ -1145,26 +1149,474 @@ void Inventory_ChangeAmmo(s16 item, s16 ammoChange) {
...
+    InterfaceContext* interfaceCtx = &play->interfaceCtx;
+    s16 magicBarY;
+
+    OPEN_DISPS(play->state.gfxCtx);
+
+    if (gSaveContext.save.playerData.magicLevel != 0) {
+        if (gSaveContext.save.playerData.healthCapacity > 0xA0) {
+            magicBarY = 42; // two rows of hearts
+        } else {
+            magicBarY = 34; // one row of hearts
```
Comment (EllipticEllipsis) [inline src/code/z_parameter.c:1557]: Did these end up getting defines or not?

**Review by EllipticEllipsis — COMMENTED** [review]
Well, it doesn't seem to have improved much from OoT... I don't see the part where Chateau Romani makes it blue?

File: include/z64save.h [hunk 3]
```diff
@@ -24,6 +24,36 @@ typedef enum RespawnMode {
 
 #define SAVE_BUFFER_SIZE 0x4000
 
+typedef enum {
+    /* 00 */ MAGIC_STATE_IDLE, // Regular gameplay
+    /* 01 */ MAGIC_STATE_CONSUME_SETUP, // Sets the speed in which magic border flashes
+    /* 02 */ MAGIC_STATE_CONSUME, // Consume magic until target is reached or no more magic is available
+    /* 03 */ MAGIC_STATE_METER_FLASH_1, // Flashes border
+    /* 04 */ MAGIC_STATE_METER_FLASH_2, // Flashes border and draws yellow magic to preview target consumption
```
[the remark this answers is not in the document]
Comment (engineer124) [reply include/z64save.h:32]: It wasn't no, but no better suggestions were ever discussed besides `FLASH_1`/`FLASH_2`/`FLASH_3` so it could be an option...

[the remark this answers is not in the document]
Comment (EllipticEllipsis) [reply include/z64save.h:32]: I'm probably mixing it up with the next enum down, they're not exactly the most distinct things.

File: include/z64save.h [hunk 4]
```diff
@@ -24,6 +24,36 @@ typedef enum RespawnMode {
...
+    /* 05 */ MAGIC_STATE_RESET, // Reset colors and return to idle
+    /* 06 */ MAGIC_STATE_METER_FLASH_3, // Flashes border with no additional behaviour
```
[the remark this answers is not in the document]
Comment (engineer124) [reply include/z64save.h:34]: In parameter, none at all. OoT uses `FLASH_1` to freeze dark-link and that's the only difference between `FLASH_1` and `FLASH_3`. I would say we need 100% to fully confirm there's no difference (i.e. having both 1/3 is just a remnant of OoT), but I would say it's likely just a remnant

File: src/code/z_parameter.c [hunk 5]
```diff
@@ -1145,26 +1149,474 @@ void Inventory_ChangeAmmo(s16 item, s16 ammoChange) {
...
+        if (gSaveContext.magicCapacity != 0) {
+            play_sound(NA_SE_SY_ERROR);
+        }
+        return false;
+    }
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_parameter/func_80115D5C.s")
+    switch (type) {
+        case MAGIC_CONSUME_NOW:
+        case MAGIC_CONSUME_NOW_ALT:
```
[the remark this answers is not in the document]
Comment (engineer124) [reply src/code/z_parameter.c:1191]: `MAGIC_CONSUME_NOW_ALT` goes entirely unused in all of OoT. I suspect it's the same in MM. But you never know what's lurking in player. And I think there are a few more magic actors in MM, so who knows. But very likely also unused tbh

File: src/code/z_parameter.c (hunk 2 above)
Comment (engineer124) [reply src/code/z_parameter.c:1557]: In OoT, these were REG's which naturally got defines. In MM, they dropped the REGs here and just used hard constants

Comment (engineer124) [reply src/code/z_parameter.c:1557]: They dropped a lot of REGs in parameter actually

File: src/code/z_actor.c [hunk 6]
```diff
@@ -2902,10 +2902,10 @@ void Actor_DrawAll(PlayState* play, ActorContext* actorCtx) {
     gSPDisplayList(sp58, &ref2[1]);
     POLY_XLU_DISP = &ref2[1];
 
-    if (play->actorCtx.unk3 != 0) {
+    if (play->actorCtx.lensActive) {
         Math_StepToC(&play->actorCtx.unk4, 100, 20);
```
[the remark this answers is not in the document]
Comment (engineer124) [reply src/code/z_actor.c:2906]: There's a few lens things I can name here if you'd like:
`actorCtx.unk3` -> `actorCtx.lensToggleTimer`? (or maybe just `actorCtx.lensTimer`)
`actorCtx.unkB` -> `actorCtx.lensActorsDrawn`
`func_800B9EF4` -> `Actor_DrawLensActors`
`func_800B9E84` -> `Actor_DrawLensOverlay`
`D_801DE890` -> `gLensOfTruthMaskTex`

File: include/z64save.h [hunk 7]
```diff
@@ -24,6 +24,36 @@ typedef enum RespawnMode {
...
+
+typedef enum {
+    /* 0 */ MAGIC_CONSUME_NOW, // Consume Magic immediately without preview
+    /* 1 */ MAGIC_CONSUME_WAIT_NO_PREVIEW, // Sets consume target but waits to consume. No yellow magic preview to target consumption. Unused
+    /* 2 */ MAGIC_CONSUME_NOW_ALT, // Identical behaviour to MAGIC_CONSUME_NOW. Unused
+    /* 3 */ MAGIC_CONSUME_LENS, // Lens consumption
+    /* 4 */ MAGIC_CONSUME_WAIT_PREVIEW, // Sets consume target but waits to consume. Draws yellow magic to target consumption
+    /* 5 */ MAGIC_CONSUME_GORON_ZORA, // Zora shock and Goron Spike Roll slow consumption
+    /* 6 */ MAGIC_CONSUME_GIANTS_MASK, // Giants Mask Slow consumption
+    /* 7 */ MAGIC_CONSUME_DEITY_BEAM // Fierce Deity Beam consumption, consumed magic now and not via request
```
[the remark this answers is not in the document]
Comment (engineer124) [reply include/z64save.h:51]: Initially I like `DEITY_BEAM` a little more than `SWORD_BEAM` but yeah should have a discussion about it.

File: src/code/z_parameter.c (hunk 2 above)
Comment (EllipticEllipsis) [reply src/code/z_parameter.c:1560]: I think it wouldn't be a bad idea to have defines for all the numbers that were regs: it makes it easier to move code between, and it's clearer what they're for than having to comment on all of them (plus enables arithmetic if appropriate).

File: src/code/z_actor.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (EllipticEllipsis) [reply src/code/z_actor.c]: Yeah, provided we're clear on what's handled differently from OoT here, I think that's a good idea. But maybe not for this PR?

File: include/z64save.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (EllipticEllipsis) [reply include/z64save.h:51]: I'm a bit in two minds here, I'd want to understand how it's handled before committing.

File: src/code/z_actor.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (engineer124) [reply src/code/z_actor.c]: afaik, the actual overall purpose of all these functions are the same, it just has slightly different implementation:

- I confirmed `gLensOfTruthMaskTex` was the same as OoT with Texture64
-  `Actor_DrawLensOverlay` is definitely drawing the purple circle, but goes through the fbdemo system instead of drawing it directly
- `Actor_DrawLensActors` is definitely drawing lens actors, but the implementation is different
- `actorCtx.lensActorsDrawn` seems like an accurate name
- You're original comment, `actorCtx.unk4` -> `actorCtx.lensMaskRadius` is actually a bad name because while this variable does control the size of the lens circle drawn on screen, it's proportional to `1 - unk4`, i.e. a value close to 0 will draw the circle at it's widest, and a value of 100 will draw the circle at a smaller radius. So it does relate to the size of the circle, but inversely (and not multiplicatively inverse, but through the additive inverse if there's a word for that)

[the remark this answers is not in the document]
Comment (EllipticEllipsis) [reply src/code/z_actor.c]: Maybe something like "fully on proportion", but actually good.

File: src/overlays/actors/ovl_En_Elf/z_en_elf.c [hunk 8]
```diff
@@ -675,8 +675,8 @@ void func_8088DD34(EnElf* this, PlayState* play) {
         !func_8088C804(&this->actor.world.pos, &refActor->actor.world.pos, 10.0f)) {
         Health_ChangeBy(play, 0x80);
         if (this->fairyFlags & 0x200) {
-            Parameter_AddMagic(play, ((void)0, gSaveContext.unk_3F30) +
-                                         (gSaveContext.save.playerData.doubleMagic * 0x30) + 0x30);
+            Magic_Add(play, ((void)0, gSaveContext.magicFillTarget) +
+                                (gSaveContext.save.playerData.isDoubleMagicAcquired + 1) * MAGIC_NORMAL_METER);
```
Comment (engineer124) [inline src/overlays/actors/ovl_En_Elf/z_en_elf.c:679]: Just documenting this here for reference, maybe this line deserves a macro:

This line shows up over and over again and seems equivalent to OoT's `Magic_Fill(play)` function, which was removed in MM. So if MM want's to fill the magic meter to full, this line here how it's called.

The `(gSaveContext.save.playerData.isDoubleMagicAcquired + 1) * MAGIC_NORMAL_METER` part should already be equal to the current magic capacity. So 
`Magic_Add((gSaveContext.save.playerData.isDoubleMagicAcquired + 1) * MAGIC_NORMAL_METER` 
alone should already add enough magic that it's guaranteed to fill to capacity. 

Adding on `magicFillTarget` to that is strange as this doesn't really do anything as adding more to a already full magic refill will just lead to a full magic refill, and it's unclear as to the logic as to what `magicFillTarget` is expected to be when calling this function in normal gameplay.

File: include/variables.h [hunk 9]
```diff
@@ -2742,7 +2742,7 @@ extern f32 D_801DE860;
 extern f32 D_801DE864;
 extern f32 D_801DE868;
 extern f32 D_801DE884;
-extern TexturePtr D_801DE890[];
+extern TexturePtr gLensOfTruthMaskTex[];
```
Comment (hensldm) [inline include/variables.h:2745]: This doesn't seem like an array of `TexturePtr`s? Especially with the name you are giving it.

File: src/overlays/actors/ovl_En_Elf/z_en_elf.c (hunk 8 above)
Comment (hensldm) [reply src/overlays/actors/ovl_En_Elf/z_en_elf.c:679]: I would tend to agree. Looks like it is used often enough in this PR alone.

**Review by hensldm — APPROVED** [review]

File: src/code/z_actor.c [hunk 10]
```diff
@@ -2606,8 +2606,8 @@ s32 Actor_RecordUndrawnActor(PlayState* play, Actor* actor) {
     return true;
 }
 
-void Actor_DrawLensOverlay(Gfx** gfxP, s32 lensMaskRadius) {
-    func_80164C14(gfxP, gLensOfTruthMaskTex, 4, 0, 6, 6, ((100 - lensMaskRadius) * 0.003f) + 1.0f);
+void Actor_DrawLensOverlay(Gfx** gfxP, s32 lensMaskSize) {
+    func_80164C14(gfxP, &gLensOfTruthMaskTex, 4, 0, 6, 6, ((100 - lensMaskSize) * 0.003f) + 1.0f);
```
Comment (EllipticEllipsis) [inline src/code/z_actor.c:2610]: This 100 is the LENS_SIZE 100?

File: src/overlays/actors/ovl_En_Elf/z_en_elf.c [hunk 11]
```diff
@@ -675,8 +675,7 @@ void func_8088DD34(EnElf* this, PlayState* play) {
         !func_8088C804(&this->actor.world.pos, &refActor->actor.world.pos, 10.0f)) {
         Health_ChangeBy(play, 0x80);
         if (this->fairyFlags & 0x200) {
-            Magic_Add(play, ((void)0, gSaveContext.magicFillTarget) +
-                                (gSaveContext.save.playerData.isDoubleMagicAcquired + 1) * MAGIC_NORMAL_METER);
+            Magic_Add(play, MAGIC_ADD_TO_CAPACITY);
```
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_En_Elf/z_en_elf.c:678]: I don't like the wording here, it could mean "FILL_TO_CAPACITY" or "ADD_ON_TOP_OF_CAPACITY", and the former is what it should be, right?

File: tools/namefixer.py [hunk 12]
```diff
@@ -695,10 +695,17 @@
...
     "play->actorCtx.unk3": "play->actorCtx.lensActive",
 
+    "play->sceneLoadFlag": "play->transitionTrigger",
+    "play->unk_18B4A": "play->transitionMode",
+    "play->unk_1887F": "play->transitionType",
+    "gSaveContext.nextTransition": "gSaveContext.nextTransitionType",
+    "gSaveContext.unk_3F48": "gSaveContext.cutsceneTransitionControl",
+    "gSaveContext.fadeDuration": "gSaveContext.transFadeDuration",
+    "gSaveContext.fadeSpeed": "gSaveContext.transWipeSpeed",
+
```
Comment (EllipticEllipsis) [inline tools/namefixer.py:708]: Probably need the lens stuff in here too.

File: src/overlays/actors/ovl_En_Elf/z_en_elf.c (hunk 11 above)
Comment (engineer124) [reply src/overlays/actors/ovl_En_Elf/z_en_elf.c:678]: It should be `FILL_TO_CAPACITY` but the reason I avoided that was to indicate that this should be for the `Magic_Add` function, maybe `MAGIC_ADD_TO_FILL_CAPACITY`? Although that sounds weird. Maybe `MAGIC_FILL_TO_CAPACITY` is fine?

Comment (EllipticEllipsis) [reply src/overlays/actors/ovl_En_Elf/z_en_elf.c:678]: I think just commenting above the macro will have to do, it's very difficult to explain that it's filling the bar to capacity rather than adding capacity in some way otherwise.

File: include/z64save.h [hunk 13]
```diff
@@ -105,13 +135,13 @@ typedef struct SavePlayerData {
     /* 0x0008 */ char playerName[8];                    // "player_name"
     /* 0x0010 */ s16 healthCapacity;                    // "max_life"
     /* 0x0012 */ s16 health;                            // "now_life"
-    /* 0x0014 */ s8 magicLevel;                         // "magic_max"
-    /* 0x0015 */ s8 magic;                              // "magic_now"
+    /* 0x0014 */ s8 magicLevel; // 0 for no magic/new load, 1 for magic, 2 for double magic "magic_max"
```
Comment (tom-overton) [inline include/z64save.h:138]: These original names look kind of weird now. Before, they were nicely aligned with each other. Now, the ones with comments are shifted pretty far to the side, and there's no punctuation (comma, period, semicolon, etc) to signal that the comment is over and the original name starts here. I'm not sure how best to handle this sort of thing, but it's only going to get worse as we document more of these fields. Maybe something like `...2 for double magic. Originally "magic_max"`? Though that wouldn't fix the alignment thing.

File: src/code/z_parameter.c [hunk 14]
```diff
@@ -1145,26 +1149,484 @@ void Inventory_ChangeAmmo(s16 item, s16 ammoChange) {
...
+            // Consume magic immediately
+            // Ex. Deku Bubble
+            if ((gSaveContext.magicState == MAGIC_STATE_IDLE) ||
+                (gSaveContext.magicState == MAGIC_STATE_CONSUME_LENS)) {
+                if (gSaveContext.magicState == MAGIC_STATE_CONSUME_LENS) {
+                    play->actorCtx.lensActive = false;
+                }
+                if (gSaveContext.save.weekEventReg[14] & 8) {
+                    // Consumed Chateau Romani
```
Comment (tom-overton) [inline src/code/z_parameter.c:1200]: I know Anghelo has a PR out for doing some weekEventRegs defines, but it might make sense to make a define for this in this PR, since it pops up so often. I also have a minor wording quibble here; this whole function uses the word "consume" a ton, so I'd prefer "Drank Chateau Romani" or something just to make it clear that no, this is actually not related to magic consumption.

File: include/z64save.h [hunk 15]
```diff
@@ -24,6 +24,36 @@ typedef enum RespawnMode {
...
+    /* 2 */ MAGIC_CONSUME_NOW_ALT, // Identical behaviour to MAGIC_CONSUME_NOW. Unused
+    /* 3 */ MAGIC_CONSUME_LENS, // Lens consumption
+    /* 4 */ MAGIC_CONSUME_WAIT_PREVIEW, // Sets consume target but waits to consume. Draws yellow magic to target consumption
+    /* 5 */ MAGIC_CONSUME_GORON_ZORA, // Zora shock and Goron Spike Roll slow consumption
+    /* 6 */ MAGIC_CONSUME_GIANTS_MASK, // Giants Mask Slow consumption
+    /* 7 */ MAGIC_CONSUME_DEITY_BEAM // Fierce Deity Beam consumption, consumed magic now and not via request
+} MagicChangeType;
+
+#define MAGIC_NORMAL_METER 0x30
+#define MAGIC_DOUBLE_METER (2 * MAGIC_NORMAL_METER)
```
Comment (tom-overton) [inline include/z64save.h:55]: I hope I'm not stumbling into a conversation that was already settled a long time ago, but I prefer magic to be in decimal rather than hexadecimal. The amount of magic that a magic arrow consumes is directly presented to the player:
NKZE01_2022-07-25_01-47-44

If a unit of magic was just some weird internal thing that players weren't really aware of, I'd be totally fine keeping it as hex. But since the game actually just straight up says "a Fire Arrow consumes two units of magic", the player *could* figure out how much magic they have at any given point just by counting how many Fire Arrows they can shoot off before they run out of magic. And given that it's countable, I'd prefer decimal in this case.

Comment (EllipticEllipsis) [reply include/z64save.h:55]: There is a significant problem with that, namely that the internal magic number is *not* the MP amount the game tells the player, it's actually double. To use your example, OoT Player has
```c
static u8 sMagicArrowCosts[] = { 4, 4, 8 };
```
Lens is so far the only thing that works with single numbers in this system.

Comment (engineer124) [reply include/z64save.h:55]: I did a test in MM to confirm these numbers, and yes: the numbers are the same as OoT.

i.e.
(MP2) = 4 units of magic in-game
(MP4) = 8 units of magic in-game

I don't necessarily have a preference to hex vs dec for magic. I don't see this MP2 being a reason to push dec over hex as it's just a single digit value that's half the in-game value of magic

The only real hard-coded magic values to begin with is:
`#define MAGIC_NORMAL_METER 0x30` for half a magic bar 
`gSaveContext.save.playerData.magic += 0x10;` for filling and extending the magic bar
`gSaveContext.save.playerData.magic--;` for some types of magic consumption
`gSaveContext.save.playerData.magic -= 2;` for other types of magic consumption
`gSaveContext.save.playerData.magic += 4;` for adding magic

Note that I've defined both double magic and magic jar refills in terms of `MAGIC_NORMAL_METER`

So all this question really begs is should we use
`#define MAGIC_NORMAL_METER 48`
`gSaveContext.save.playerData.magic += 16;`
instead, which I don't have a preference. Hex is slightly more even, and aligns with health, but it would only change two lines.

File: src/code/z_parameter.c (hunk 14 above)
Comment (engineer124) [reply src/code/z_parameter.c:1200]: Since there are many more parameter PRs coming, I think I'd prefer to name the weekeventreg once the Anghelo PR is in and avoid that conflict (I can do it nicely when I PR `Interface_Update` which has the remaining code on magic). I can change `consume` -> `drank` though (it's not my fault Link consumes a lot of various things on his adventures haha)

File: include/z64save.h (hunk 15 above)
Comment (EllipticEllipsis) [reply include/z64save.h:55]: MM doesn't have any big magic-using items. OoT has Nayru's Love, which uses 24 ("MP12")
```c
static u8 sMagicSpellCosts[] = { 12, 24, 24, 12, 24, 12 };
```
I suspect that the actual numbers were added as an afterthought, I don't think any other Zelda game gives actual numbers for magic use. I suspect this is going to be another "make a macro for external use" thing, which I think I mentioned before for hearts.

Comment (tom-overton) [reply include/z64save.h:55]: Yeah I didn't realize what was presented to the user was any different from the internal representation. I'm fine with using hex then.

File: include/z64save.h (hunk 13 above)
Comment (engineer124) [reply include/z64save.h]: These struct members will need to be documented eventually, just having only the original names was more of a stepping stone. see `z64actor.h` (specifically `Actor`) which is better documented, and that's presumably how these structs will end up.

File: include/z64save.h [hunk 16]
```diff
@@ -24,6 +24,36 @@ typedef enum RespawnMode {
...
+    /* 2  */ MAGIC_STATE_CONSUME, // Consume magic until target is reached or no more magic is available
+    /* 3  */ MAGIC_STATE_METER_FLASH_1, // Flashes border
+    /* 4  */ MAGIC_STATE_METER_FLASH_2, // Flashes border and draws yellow magic to preview target consumption
+    /* 5  */ MAGIC_STATE_RESET, // Reset colors and return to idle
+    /* 6  */ MAGIC_STATE_METER_FLASH_3, // Flashes border with no additional behaviour
+    /* 7  */ MAGIC_STATE_CONSUME_LENS, // Magic slowly consumed by lens
+    /* 8  */ MAGIC_STATE_STEP_CAPACITY, // Step `magicCapacity` to full capacity
+    /* 9  */ MAGIC_STATE_FILL, // Add magic until magicFillTarget is reached
+    /* 10 */ MAGIC_STATE_CONSUME_GORON_ZORA_SETUP,
+    /* 11 */ MAGIC_STATE_CONSUME_GORON_ZORA, // Magic slowly consumed by goron spikes or zora shocks
```
Comment (EllipticEllipsis) [inline include/z64save.h:39]:
```suggestion
    /* 11 */ MAGIC_STATE_CONSUME_GORON_ZORA, // Magic slowly consumed by Goron spiked rolling or Zora electric shield
```

File: include/z64save.h [hunk 17]
```diff
@@ -24,6 +24,36 @@ typedef enum RespawnMode {
...
 
+typedef enum {
+    /* 0  */ MAGIC_STATE_IDLE, // Regular gameplay
+    /* 1  */ MAGIC_STATE_CONSUME_SETUP, // Sets the speed at which the magic border flashes
+    /* 2  */ MAGIC_STATE_CONSUME, // Consume magic until target is reached or no more magic is available
+    /* 3  */ MAGIC_STATE_METER_FLASH_1, // Flashes border
+    /* 4  */ MAGIC_STATE_METER_FLASH_2, // Flashes border and draws yellow magic to preview target consumption
+    /* 5  */ MAGIC_STATE_RESET, // Reset colors and return to idle
+    /* 6  */ MAGIC_STATE_METER_FLASH_3, // Flashes border with no additional behaviour
+    /* 7  */ MAGIC_STATE_CONSUME_LENS, // Magic slowly consumed by lens
```
Comment (EllipticEllipsis) [inline include/z64save.h:35]:
```suggestion
    /* 7  */ MAGIC_STATE_CONSUME_LENS, // Magic slowly consumed by Lens of Truth
```

File: include/z64save.h [hunk 18]
```diff
@@ -24,6 +24,36 @@ typedef enum RespawnMode {
...
+    /* 8  */ MAGIC_STATE_STEP_CAPACITY, // Step `magicCapacity` to full capacity
+    /* 9  */ MAGIC_STATE_FILL, // Add magic until magicFillTarget is reached
+    /* 10 */ MAGIC_STATE_CONSUME_GORON_ZORA_SETUP,
+    /* 11 */ MAGIC_STATE_CONSUME_GORON_ZORA, // Magic slowly consumed by goron spikes or zora shocks
+    /* 12 */ MAGIC_STATE_CONSUME_GIANTS_MASK // Magic slowly consumed by giants mask
```
Comment (EllipticEllipsis) [inline include/z64save.h:40]:
```suggestion
    /* 12 */ MAGIC_STATE_CONSUME_GIANTS_MASK // Magic slowly consumed by Giant's Mask
```

File: include/z64save.h [hunk 19]
```diff
@@ -24,6 +24,36 @@ typedef enum RespawnMode {
...
+} MagicState;
+
+typedef enum {
+    /* 0 */ MAGIC_CONSUME_NOW, // Consume Magic immediately without preview
```
Comment (EllipticEllipsis) [inline include/z64save.h:44]: Decide whether "magic" is capitalised or not, it's inconsistent in the docs here.

File: include/z64save.h [hunk 20]
```diff
@@ -24,6 +24,36 @@ typedef enum RespawnMode {
...
+    /* 1 */ MAGIC_CONSUME_WAIT_NO_PREVIEW, // Sets consume target but waits to consume. No yellow magic preview to target consumption. Unused
+    /* 2 */ MAGIC_CONSUME_NOW_ALT, // Identical behaviour to MAGIC_CONSUME_NOW. Unused
+    /* 3 */ MAGIC_CONSUME_LENS, // Lens consumption
```
Comment (EllipticEllipsis) [inline include/z64save.h:47]:
```suggestion
    /* 3 */ MAGIC_CONSUME_LENS, // Lens of Truth consumption
```
Generally there's no reason to be overly brief in docs like this.

File: include/z64save.h [hunk 21]
```diff
@@ -24,6 +24,36 @@ typedef enum RespawnMode {
...
+    /* 4 */ MAGIC_CONSUME_WAIT_PREVIEW, // Sets consume target but waits to consume. Draws yellow magic to target consumption
```
Comment (EllipticEllipsis) [inline include/z64save.h:48]:
```suggestion
    /* 4 */ MAGIC_CONSUME_WAIT_PREVIEW, // Sets consume target but waits to consume. Show magic to be consumed in yellow.
```

File: include/z64save.h [hunk 22]
```diff
@@ -24,6 +24,36 @@ typedef enum RespawnMode {
...
+    /* 5 */ MAGIC_CONSUME_GORON_ZORA, // Zora shock and Goron Spike Roll slow consumption
```
Comment (EllipticEllipsis) [inline include/z64save.h:49]:
```suggestion
    /* 5 */ MAGIC_CONSUME_GORON_ZORA, // Zora electric barrier and Goron spiked rolling slow consumption
```
Or at least make the terminology consistent.

File: include/z64save.h [hunk 23]
```diff
@@ -24,6 +24,36 @@ typedef enum RespawnMode {
...
+    /* 6 */ MAGIC_CONSUME_GIANTS_MASK, // Giants Mask Slow consumption
```
Comment (EllipticEllipsis) [inline include/z64save.h:50]:
```suggestion
    /* 6 */ MAGIC_CONSUME_GIANTS_MASK, // Giant's Mask slow consumption
```
Presumably?

File: include/z64save.h [hunk 24]
```diff
@@ -226,13 +256,13 @@ typedef struct SaveContext {
...
-    /* 0x3F28 */ s16 unk_3F28;                          // "magic_flag"
-    /* 0x3F2A */ s16 unk_3F2A;                          // "recovery_magic_flag"
-    /* 0x3F2C */ s16 unk_3F2C;                          // "keep_magic_flag"
-    /* 0x3F2E */ s16 unk_3F2E;                          // "magic_now_max"
-    /* 0x3F30 */ s16 unk_3F30;                          // "magic_now_now"
-    /* 0x3F32 */ s16 unk_3F32;                          // "magic_used"
-    /* 0x3F34 */ s16 unk_3F34;                          // "magic_recovery"
+    /* 0x3F28 */ s16 magicState; // determines magic meter behavior on each frame "magic_flag"
+    /* 0x3F2A */ s16 isMagicRequested; // a request to add magic has been given "recovery_magic_flag"
+    /* 0x3F2C */ s16 magicFlag; // Set to 0 in func_80812D94, otherwise unused "keep_magic_flag"
```
Comment (EllipticEllipsis) [inline include/z64save.h:261]:
```suggestion
    /* 0x3F2C */ s16 magicFlag; // Set to 0 in func_80812D94(), otherwise unused "keep_magic_flag"
```

File: src/code/z_parameter.c [hunk 25]
```diff
@@ -600,15 +604,15 @@ u8 Item_Give(PlayState* play, u8 item) {
         return item;
 
     } else if (item == ITEM_MAGIC_SMALL) {
-        Parameter_AddMagic(play, 0x18);
+        Magic_Add(play, MAGIC_NORMAL_METER / 2);
```
Comment (EllipticEllipsis) [inline src/code/z_parameter.c:607]: I wouldn't necessarily say that the magic recovered from a magic jar is inherently related to the size of the meter.

File: src/code/z_parameter.c [hunk 26]
```diff
@@ -1145,26 +1149,484 @@ void Inventory_ChangeAmmo(s16 item, s16 ammoChange) {
...
+            play_sound(NA_SE_SY_ERROR);
+        }
+        return false;
+    }
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_parameter/func_80115D5C.s")
+    switch (type) {
+        case MAGIC_CONSUME_NOW:
+        case MAGIC_CONSUME_NOW_ALT:
+            // Drank magic immediately
```
Comment (EllipticEllipsis) [inline src/code/z_parameter.c:1192]:
```suggestion
            // Drain magic immediately
```
?

File: src/code/z_parameter.c [hunk 27]
```diff
@@ -1145,26 +1149,484 @@ void Inventory_ChangeAmmo(s16 item, s16 ammoChange) {
...
+            // Ex. Deku Bubble
```
Comment (EllipticEllipsis) [inline src/code/z_parameter.c:1193]:
```suggestion
            // e.g. Deku Bubble
```
and not really any need to be on its own line imo.

File: src/code/z_parameter.c [hunk 28]
```diff
@@ -1145,26 +1149,484 @@ void Inventory_ChangeAmmo(s16 item, s16 ammoChange) {
...
+                return true;
+            } else {
+                return false;
+            }
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_parameter/func_80116088.s")
+        case MAGIC_CONSUME_WAIT_PREVIEW:
+            // Sets consume target but waits to consume.
+            // Preview consumption with a yellow bar
+            // Ex. Spin Attack
```
Comment (EllipticEllipsis) [inline src/code/z_parameter.c:1249]:
```suggestion
            // e.g. Spin Attack
```

File: src/code/z_parameter.c [hunk 29]
```diff
@@ -1145,26 +1149,484 @@ void Inventory_ChangeAmmo(s16 item, s16 ammoChange) {
...
+                gSaveContext.magicState = MAGIC_STATE_METER_FLASH_2;
+                return true;
+            } else {
+                play_sound(NA_SE_SY_ERROR);
+                return false;
+            }
 
-s16 magicBorderColors[][3] = {
+        case MAGIC_CONSUME_GORON_ZORA:
+            // Zora Shock, Goron Spike Roll
```
Comment (EllipticEllipsis) [inline src/code/z_parameter.c:1264]:
```suggestion
            // Zora electric shield, Goron spike roll
```
etc.

File: src/code/z_parameter.c [hunk 30]
```diff
@@ -1145,26 +1149,484 @@ void Inventory_ChangeAmmo(s16 item, s16 ammoChange) {
...
+            if (gSaveContext.save.playerData.magic != 0) {
+                interfaceCtx->magicConsumptionTimer = 10;
+                gSaveContext.magicState = MAGIC_STATE_CONSUME_GORON_ZORA_SETUP;
+                return true;
+            } else {
+                return false;
+            }
+
+        case MAGIC_CONSUME_GIANTS_MASK:
+            // Wearing Giants Mask
```
Comment (EllipticEllipsis) [inline src/code/z_parameter.c:1274]:
```suggestion
            // Wearing Giant's Mask
```

File: src/code/z_parameter.c [hunk 31]
```diff
@@ -1145,26 +1149,484 @@ void Inventory_ChangeAmmo(s16 item, s16 ammoChange) {
...
+            }
+            break;
+
+        case MAGIC_STATE_RESET:
+            sMagicMeterOutlinePrimRed = sMagicMeterOutlinePrimGreen = sMagicMeterOutlinePrimBlue = 255;
+            gSaveContext.magicState = MAGIC_STATE_IDLE;
+            break;
+
+        case MAGIC_STATE_CONSUME_LENS:
+            // Slowly consume magic while lens is on
```
Comment (EllipticEllipsis) [inline src/code/z_parameter.c:1461]:
```suggestion
            // Slowly consume magic while Lens of Truth is active
```

File: src/code/z_parameter.c [hunk 32]
```diff
@@ -1145,26 +1149,484 @@ void Inventory_ChangeAmmo(s16 item, s16 ammoChange) {
...
+                (play->transitionTrigger == TRANS_TRIGGER_OFF) && (play->transitionMode == TRANS_MODE_OFF) &&
+                !Play_InCsMode(play)) {
+
+                if ((gSaveContext.save.playerData.magic == 0) ||
+                    ((func_801242DC(play) >= 2) && (func_801242DC(play) <= 4)) ||
+                    ((BUTTON_ITEM_EQUIP(0, EQUIP_SLOT_C_LEFT) != ITEM_LENS) &&
+                     (BUTTON_ITEM_EQUIP(0, EQUIP_SLOT_C_DOWN) != ITEM_LENS) &&
+                     (BUTTON_ITEM_EQUIP(0, EQUIP_SLOT_C_RIGHT) != ITEM_LENS)) ||
+                    !play->actorCtx.lensActive) {
+                    // Force lens off and set magic state to idle
```
Comment (EllipticEllipsis) [inline src/code/z_parameter.c:1473]:
```suggestion
                    // Deactivate Lens of Truth and set magic state to idle
```

**Review by EllipticEllipsis — COMMENTED** [review]
Last few things, almost entirely to try and get the docs consistent.

File: include/z64save.h (hunk 16 above)
Comment (engineer124) [reply include/z64save.h:39]: Would you call it a shield? When standing, shield makes sense. But you can use this while swimming as an attack, which doesn't feel like a shield

File: include/z64save.h (hunk 22 above)
Comment (engineer124) [reply include/z64save.h:49]: Ah. I like barrier. I'll stick with that

File: src/code/z_parameter.c (hunk 25 above)
Comment (engineer124) [reply src/code/z_parameter.c:607]: It's a little subjective, you could design magic to restore a fixed amount or a fraction of the total. At least with this interpretation, it saves an extra define and gives scale to how much magic is restored

**Review by EllipticEllipsis — APPROVED** [review]

File: include/z64save.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (Kenix3) [reply include/z64save.h:51]: I like Deity Beam over Sword Beam personally.

File: src/code/z_parameter.c (hunk 25 above)
Comment (Kenix3) [reply src/code/z_parameter.c:607]: While I do agree it's not inherently tied to the size of the magic meter, I'd say it's clear that chose that value because it's half. I also think we should keep it this way because it gives someone reading the code context on how much magic "0"18" is restoring, in other words documentation. :)

**Review by Kenix3 — APPROVED** [review]
